aptos-move/e2e-move-tests/tests/stake.rs

```diff
@@ -0,0 +1,184 @@
+    let operator = harness.new_account_at(AccountAddress::from_hex_literal("0x234").unwrap());
+    let owner_address = *owner.address();
+    let operator_address = *operator.address();
+
+    // Initialize and add stake.
+    let stake_amount = 100_000_000;
+    assert_success!(harness.initialize_staking(
+        &owner,
+        stake_amount,
+        operator_address,
+        owner_address
+    ));
+    let stake_pool = harness.get_stake_pool(&owner_address);
+    assert_eq!(stake_pool.active, stake_amount);
+    assert_eq!(stake_pool.operator_address, operator_address);
+    assert_eq!(stake_pool.delegated_voter, owner_address);
+
+    // Join the validator set.
+    assert_success!(harness.rotate_consensus_key(&operator, owner_address));
+    assert_success!(harness.join_validator_set(&operator, owner_address));
+    harness.new_epoch();
+
+    // Validator should now be locked up.
+    let stake_pool = harness.get_stake_pool(&owner_address);
+    assert_eq!(
+        stake_pool.locked_until_secs,
+        harness.executor.get_block_time_seconds() + 7200
+    );
+
+    // Unlock stake.
+    assert_success!(harness.unlock_stake(&owner, stake_amount / 2));
+    let stake_pool = harness.get_stake_pool(&owner_address);
+    assert_eq!(stake_pool.active, stake_amount / 2);
+    assert_eq!(stake_pool.pending_inactive, stake_amount / 2);
+
+    // Wait until stake is unlocked.
+    harness.fast_forward(7200);
+    harness.new_epoch();
+    let stake_pool = harness.get_stake_pool(&owner_address);
+    assert_eq!(stake_pool.pending_inactive, 0);
+    assert_eq!(stake_pool.inactive, stake_amount / 2);
+
+    // Withdraw and verify that coins are returned.
+    assert_success!(harness.withdraw_stake(&owner, stake_amount / 2));
+    let stake_pool = harness.get_stake_pool(&owner_address);
+    assert_eq!(stake_pool.inactive, 0);
+
+    // Verify that lockup has been renewed for remaining stake.
+    assert_eq!(stake_pool.active, stake_amount / 2);
+    assert_eq!(
+        stake_pool.locked_until_secs,
+        harness.executor.get_block_time_seconds() + 7200
+    );
+
+    // Validator takes the rest of the stake out.
+    assert_success!(harness.unlock_stake(&owner, stake_amount / 2));
+    harness.fast_forward(7200);
+    harness.new_epoch();
+    assert_success!(harness.withdraw_stake(&owner, stake_amount / 2));
+    let stake_pool = harness.get_stake_pool(&owner_address);
+    assert_eq!(stake_pool.active, 0);
+    assert_eq!(stake_pool.inactive, 0);
+}
+
+#[test]
+fn test_staking_rewards() {
+    let mut harness = MoveHarness::new();
+    enable_golden!(harness);
+    let validator_1 = harness.new_account_at(AccountAddress::from_hex_literal("0x123").unwrap());
+    let validator_2 = harness.new_account_at(AccountAddress::from_hex_literal("0x234").unwrap());
+    let validator_1_address = *validator_1.address();
+    let validator_2_address = *validator_2.address();
+
+    // Initialize the validators.
+    let rewards_per_epoch = 1141;
```

Comment:
can you put a formula (or comment the formula) for this magical number. (i.e. 0.1/epochs_in_a_year, etc)

Reply:
It's hard to get the formula as the configs (rewards rate, epoch duration, etc.) are buried deep in the genesis configurations. It'll take some code to dig it out. We can probably do this in a separate PR later.